src/ir/module-utils.h

```diff
@@ -414,7 +414,16 @@ collectSignatures(Module& wasm,
       Counts& counts;
 
       TypeCounter(Counts& counts) : counts(counts) {}
+
       void visitExpression(Expression* curr) {
+        // A signature type could be anywhere.
+        if (curr->type.isRef()) {
+          auto heapType = curr->type.getHeapType();
+          if (heapType.isSignature()) {
+            counts[heapType.getSignature()]++;
+          }
+        }
+        // Specific places also have a special signature.
         if (auto* call = curr->dynCast<CallIndirect>()) {
           counts[call->sig]++;
```

Comment:
Yes, each param can be a typed function reference type. But I'm not sure what you mean in the second sentence - what do you mean by "collect signatures"? (I think this is the one place that collects them, so there is no other place to worry about?)

Reply:
There are several places we count signatures, each of which consists of params and results, and each param or result can be a typed reference type that contains a signature inside. What I meant was, when we count signatures call `call_indirect`, block types, or loop types, ..., should we check if each param/result is in turn also a typed reference type, in case it contains a signature? Maybe I'm misunderstanding something.